client/thin-replica-client/test/replica_stream_snapshot_client_test.cpp

```diff
@@ -0,0 +1,290 @@
+  vector<unique_ptr<::grpc::Server>> servers(7);
+  std::atomic_uint64_t server_hello_count{0};
+  auto run_server = [&servers, &server_hello_count](size_t i) {
+    string server_address = "0.0.0.0:" + std::to_string(50001 + i);
+    vmware::concord::replicastatesnapshot::FakeReplicaStateSnapshotService service;
+    ::grpc::ServerBuilder builder;
+    // Listen on the given address without any authentication mechanism.
+    builder.AddListeningPort(server_address, grpc::InsecureServerCredentials());
+    // Register "service" as the instance through which we'll communicate with
+    // clients. In this case it corresponds to an *synchronous* service.
+    builder.RegisterService(&service);
+    builder.SetMaxReceiveMessageSize(10 * 1024 * 1024);
+    builder.SetMaxSendMessageSize(10 * 1024 * 1024);
+    // Finally assemble the server.
+    servers[i] = builder.BuildAndStart();
+    ++server_hello_count;
+    // Wait for the server to shutdown. Note that some other thread must be
+    // responsible for shutting down the server for this call to ever return.
+    (servers[i])->Wait();
+  };
+  ThreadPool server_thread_pool{7};
+  vector<std::future<void>> server_results;
+  for (size_t i = 0; i < 7; ++i) {
+    server_results.push_back(server_thread_pool.async(run_server, i));
+  }
+
+  while (true) {
+    if (server_hello_count.load() == 7) break;
+  }
+
+  sleep(1);
+
+  vector<shared_ptr<GrpcConnection>> grpc_connections;
+  getGrpcConnections(false, grpc_connections, 7);
+  ThreadPool thread_pool{10};
+  auto read_snapshot = [&grpc_connections](size_t len) {
+    auto rss_config = std::make_unique<ReplicaStateSnapshotClientConfig>(grpc_connections, 8);
+    auto rss = std::make_unique<ReplicaStreamSnapshotClient>(std::move(rss_config));
+
+    std::shared_ptr<UpdateQueue> remote_queue = std::make_shared<BasicUpdateQueue>();
+    ASSERT_EQ(remote_queue->size(), 0);
+    ::client::replica_state_snapshot_client::SnapshotRequest rss_request;
+    rss_request.snapshot_id = len;
+    rss_request.last_received_key = "";
+    rss->readSnapshotStream(rss_request, remote_queue);
+    ThreadPool read_thread_pool{1};
+    read_thread_pool.async(
+        [&remote_queue](size_t l) {
+          size_t num_received = 0;
+          while (true) {
+            std::unique_ptr<RemoteRecepient> update;
+            try {
+              update = remote_queue->pop();
+            } catch (const concord::client::concordclient::UpdateNotFound& e) {
+              break;
+            } catch (const concord::client::concordclient::OutOfRangeSubscriptionRequest& e) {
+              break;
+            } catch (const concord::client::concordclient::InternalError& e) {
+              break;
+            } catch (const concord::client::concordclient::StreamUnavailable& e) {
+              break;
+            } catch (const concord::client::concordclient::EndOfStream& e) {
+              break;
+            } catch (...) {
+              break;
+            }
```

Comment:
Do we need to explicitly handle all types?

Reply:
How can it know which exceptions are about to be thrown? It seems like a lot of boilerplate to me, would be nice to avoid it , IMO.